### PR TITLE
[16.0][IMP] rma_sale: operation required

### LINKED ISSUE #404: RMA - require operation
### Make the rma operation required

As multiple rma operations (return, refund, replace...) are available, I find it important that the user explicitly choose the operation for each product

In rma:
- [ ] Keep the field optional on the rma but require a value when confirming the rma (by adding to _ensure_required_fields) as this dictates what to do on confirmation

In rma_sale:
- [ ] In the SO create rma wizard, add before the list of products an optional field to set the default operation for all the lines. Make on all product lines the operation computed stored based on that default operation
- [ ] In the SO create rma wizard, on each product line, make the field required if qty > 0 as this wizard confirms the rma which require an operation to be set

### PR DESCRIPTION
**Pull request by sbejaoui**
The rma operation is required in the portal, and it makes sense for it to be required in the sales wizard as well

fixes: #404 

cc/ @pedrobaeza , @jbaudoux , @lmignon , @rousseldenis

### CODE REVIEW

**Review by rousseldenis — APPROVED** [review]
Code review. LGTM

File: rma_sale/wizard/sale_order_rma_wizard_views.xml [hunk 1]
```diff
@@ -12,6 +12,7 @@
         <field name="arch" type="xml">
             <form>
                 <group>
+                    <field name="operation_id" />
```
Comment (pedrobaeza) [inline rma_sale/wizard/sale_order_rma_wizard_views.xml:15]: Why putting it twice?

File: rma/tests/test_rma.py [hunk 2]
```diff
@@ -157,7 +162,9 @@ def test_post_init_hook(self):
 
     def test_rma_replace_pick_ship(self):
         self.warehouse.write({"delivery_steps": "pick_ship"})
-        rma = self._create_rma(self.partner, self.product, 1, self.rma_loc)
+        rma = self._create_rma(
+            self.partner, self.product, 1, self.rma_loc, self.operation
```
Comment (pedrobaeza) [inline rma/tests/test_rma.py]: Can we avoid to pass on each call the operation and change the method `_create_rma` to take `self.operation` if the value is `None`? It's for reducing all the diff in the test that makes it difficult to follow.

File: rma/wizard/stock_picking_return_views.xml [hunk 3]
```diff
@@ -14,6 +14,10 @@
                         name="create_rma"
                         attrs="{'invisible': [('picking_type_code', '!=', 'outgoing')]}"
                     />
+                    <field
+                        name="rma_operation_id"
+                        attrs="{'invisible': [('create_rma', '=', False)], 'required': [('create_rma', '=', True)]}"
```
Comment (pedrobaeza) [inline rma/wizard/stock_picking_return_views.xml]: Not sure we should require it here.

File: rma_sale/wizard/sale_order_rma_wizard_views.xml (hunk 1 above)
Comment (jbaudoux) [reply rma_sale/wizard/sale_order_rma_wizard_views.xml:15]: It's the default operation that will apply to all lines. Below is the operation for the line

Comment (pedrobaeza) [reply rma_sale/wizard/sale_order_rma_wizard_views.xml:15]: OK, understood.

File: rma/static/description/index.html [hunk 4]
```diff
@@ -522,6 +522,8 @@ <h2><a class="toc-backref" href="#toc-entry-7">Contributors</a></h2>
 <li>Chafique Delli &lt;<a class="reference external" href="mailto:chafique.delli&#64;akretion.com">chafique.delli&#64;akretion.com</a>&gt;</li>
 <li>Giovanni Serra - Ooops &lt;<a class="reference external" href="mailto:giovanni&#64;ooops404.com">giovanni&#64;ooops404.com</a>&gt;</li>
 <li>Michael Tietz (MT Software) &lt;<a class="reference external" href="mailto:mtietz&#64;mt-software.de">mtietz&#64;mt-software.de</a>&gt;</li>
+<li>Jacques-Etienne Baudoux - BCIM &lt;<a class="reference external" href="mailto:je&#64;bcim.be">je&#64;bcim.be</a>&gt;</li>
+<li>Souheil Bejaoui - ACSONE SA/NV &lt;<a class="reference external" href="mailto:souheil.bejaoui&#64;acsone.eu">souheil.bejaoui&#64;acsone.eu</a>&gt;</li>
```
Comment (jbaudoux) [inline rma/static/description/index.html:526]: Can you skip the update of auto generated files?
You can use `export SKIP=oca-gen-addon-readme`

File: rma/wizard/stock_picking_return_views.xml (hunk 3 above)
Comment (sbejaoui) [reply rma/wizard/stock_picking_return_views.xml]: The RMA created from the return wizard is confirmed right away. The user will get an error if they don't fill in the operation. So, I think it's more user-friendly to make the field required in the view

File: rma/tests/test_rma.py (hunk 2 above)
Comment (sbejaoui) [reply rma/tests/test_rma.py]: Make sense

File: rma/static/description/index.html (hunk 4 above)
Comment (sbejaoui) [reply rma/static/description/index.html:526]: Done

File: rma/tests/test_rma.py (hunk 2 above)
Comment (sbejaoui) [reply rma/tests/test_rma.py]: Done

**Review by pedrobaeza — APPROVED** [review]

**Review by jbaudoux — APPROVED** [review]
LG, Thanks

**Review by jbaudoux — COMMENTED** [review]
Can you perform the same in the stock picking return ? And add the operation in the lines.
You can also create RMAs from there

**Comment by sbejaoui** [comment]
> Can you perform the same in the stock picking return ? And add the operation in the lines. You can also create RMAs from there

done

image

**Comment by sbejaoui** [comment]
@pedrobaeza , ready for merge

File: rma/wizard/stock_picking_return_views.xml [hunk 5]
```diff
@@ -8,12 +8,22 @@
         <field name="model">stock.return.picking</field>
         <field name="inherit_id" ref="stock.view_stock_return_picking_form" />
         <field name="arch" type="xml">
-            <field name="product_return_moves" position="after">
+            <xpath expr="//field[@name='product_return_moves']//tree" position="inside">
+                <field
+                    name="rma_operation_id"
+                    attrs="{'column_invisible': [('parent.create_rma', '=', False)], 'required': [('parent.create_rma', '=', True)]}"
```
Comment (jbaudoux) [inline rma/wizard/stock_picking_return_views.xml]: Like you did on sales, rma operation is only required if you return that line
```suggestion
                    attrs="{'column_invisible': [('parent.create_rma', '=', False)], 'required': [('parent.create_rma', '=', True), ('quantity', '>', 0)]}"
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]
Picking return wizard to fix

**Comment by pedrobaeza** [comment]
Please squash commits together. They don't add value as separated. And remember you can do `git commit --amend` and `git push -f`, as GitHub will put the diff of such forced push, so you avoid extra CI builds for nothing.

**Comment by sbejaoui** [comment]
@jbaudoux , could you please update your review? We'd like to merge this PR since many others depend on it.

**Review by jbaudoux — APPROVED** [review]

**Pull request merged**